### PR TITLE
Improve the performance of rules inheriting from DoNotCallMethodsBase

### LINKED ISSUE #3854: Improve performance for S3869, S3889, S3902, S3885, S1147, S1215, S3971, S4040
Do early AST-based name filtering in DoNotCallMethodsBase 

We can improve the implementation of DoNotCallMethodsBase by moving some name checks earlier. We have some symbol-based filtering in `FindDisallowedMethodSignature` , which could be done AST-based before we call `IsInValidContext`.

### LINKED ISSUE #4349: Fix perf regression in S3971 DoNotCallGCSuppressFinalize
See sonar-dotnet [Performance analysis (28 April 2021)](https://xtranet-sonarsource.atlassian.net/wiki/spaces/LANG/pages/1643610461/sonar-dotnet+Performance+analysis+28+April+2021).

DoNotCallGCSuppressFinalize - doubled the time from 30 seconds to 60 seconds on EfCore.

It might be related to  #3854 (if the fix is inside the base class, then  #3854 should be closed by this one).

### PR DESCRIPTION
**Pull request by csaba-sagi-sonarsource**
Fixes #3854
Fixes #4349

The testing of the performance included 4 runs each on 3 different projects. The sum of the times are the following:
akka 75.353 -> 21.673
efcore 116.028 -> 19.791
noda 18.35 -> 1.837

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotCallExitMethods.cs [hunk 1]
```diff
@@ -48,6 +48,7 @@ public sealed class DoNotCallExitMethods : DoNotCallMethodsCSharpBase
             !invocationSyntax
                 .Ancestors()
                 .OfType<BaseMethodDeclarationSyntax>()
+                .Where(x => x.GetIdentifierOrDefault()?.ValueText == "Main")
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotCallExitMethods.cs:51]: I was curious how this works for top-level statements and it looks like it doesn't. Shouldn't this be a false positive: https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCallExitMethods.CSharp9.cs#L3 ?

**Review by costin-zaharia-sonarsource — APPROVED** [review]
Nice! 🚀 

I think we should mark the case from https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/DoNotCallExitMethods.CSharp9.cs#L3 as a false positive and add an issue to fix it.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotCallExitMethods.cs (hunk 1 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/DoNotCallExitMethods.cs]: It was a really small change, so I fixed it in this PR. I hope you don't mind.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**